jax/interpreters/batching.py

```diff
@@ -31,51 +33,71 @@
 
 map = safe_map
 
-BatchDim = Optional[int]
-BatchDims = Sequence[BatchDim]
-@lu.transformation
-def batchfun(axis_name, axis_size, in_dims, main_type, *in_vals):
-  # analogue of `jvpfun` in ad.py
-  if axis_size is None:
-    axis_size, = {x.shape[d] for x, d in zip(in_vals, in_dims) if d is not not_mapped}
-  in_dims = in_dims() if callable(in_dims) else in_dims
-  in_dims = [canonicalize_axis(ax, np.ndim(x)) if isinstance(ax, int)
-             and not isinstance(core.get_aval(x), core.AbstractUnit)
-             else ax for x, ax in zip(in_vals, in_dims)]
-  with core.new_main(main_type, axis_name=axis_name) as main:
-    with core.extend_axis_env(axis_name, axis_size, main):
-      out_vals = yield (main, in_dims, *in_vals), {}
-      del main
-  yield out_vals
+### vmappable typeclass
+
+Vmappable = Any
+Elt = Any
+MapSpec = Any
+AxisSize = Any
+Array = Any
+ToEltHandler = Callable[[Callable, Tracer, Vmappable, MapSpec], Elt]
+FromEltHandler = Callable[[Callable, AxisSize, Elt, MapSpec], Vmappable]
+MakeIotaHandler = Callable[[AxisSize], Array]
+
+def to_elt(trace: Trace, i: Tracer, x: Vmappable, spec: MapSpec) -> Elt:
+  handler = to_elt_handlers.get(type(x))
+  if handler:
+    return handler(partial(to_elt, trace, i), i, x, spec)
+  else:
+    spec = spec and canonicalize_axis(spec, len(np.shape(x)))
+    return BatchTracer(trace, x, spec) if spec is not None else x
+to_elt_handlers: Dict[Type, ToEltHandler] = {}
+
+def from_elt(trace: Trace, axis_size: AxisSize, x: Elt, spec: MapSpec
+             ) -> Vmappable:
+  handler = from_elt_handlers.get(type(x))
+  if handler:
+    return handler(partial(from_elt, trace), axis_size, x, spec)
+  else:
+    x_ = trace.full_raise(x)
+    return matchaxis(trace.axis_name, axis_size, x_.batch_dim, spec, x_.val)
+from_elt_handlers: Dict[Type, FromEltHandler] = {}
+
+def make_iota(axis_size: AxisSize) -> Array:
+  handler = make_iota_handlers.get(type(axis_size))
+  if handler:
+    return handler(axis_size)
+  else:
+    return np.arange(axis_size)  # TODO(mattjj): make this a jax.lax.iota?
```

Comment:
A funny thing!

`vmap` now creates this `arange`/`iota`, of size equal to the axis size, every time it's called. Since we don't have any laziness built into lax anymore, unless we add a layer of laziness here (because nothing is _using_ this iota yet, though one of our dynamic shape experiments did), we'll actually end up creating this array then ignoring it. Under a `jit` that'd be DCE'd but outside a `jit` it'll actually get created.

I kept `np.arange` because I think it's the cheaper option; `np.arange(1024)` takes <1µs, whereas `lax.iota(np.int32, 1024)` seems to take ~10x that, and on other backends it'll temporarily use some device memory.

The difference is probably immaterial but I was feeling tentative about it so I decided to go with the safer option at first... I might end up adding in some laziness here if it doesn't make the API too weird.

Reply:
I ended up adding some laziness to work around an unrelated issue: jax2tf uses vmap to add an _unbounded-size_ batch axis. We had written this PR assuming there would always be an axis size _bound_ available when calling the vmapped function; after all, such a bound must exist to actually carry out a vmapped computation without requiring infinite memory. But with a staging system that allows sizes (and even their bounds) to be specified later, that information may not be available when running the vmapped computation. And jax2tf is such a staging system.

I consider this laziness to be a temporary workaround. We have a more general mechanism which could be used, namely use of the axis_size primitive, but I didn't switch to using that yet because it requires a name for the mapped axis (and so pulling on that thread might lead to several other changes...).